### PR TITLE
Fix bug in subclass unwrapper

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142405
* #141941

Let's say we have a nested TwoTensor that needs to unwrapped. Unwrapper util will create following rebuilt_stack:
```
[{elem1, elem2, subclass_2}, {elem1, elem2, subclass_1}, {elem1, elem2, nested_subclass_1}] 
```
And, we will receive following plain tensors to unflatten:
```
tensor1 (from subclass1), 
tensor2(from siubclass1), 
tensor3 (from subclass2), 
tensor4(from siubclass2)
```
When we call the first unflatten, we tried to reconstruct the second innner subclass first as it is the last one that is been flattened. The way we divide previously is:
```
todo = [tensor1, tensor2, tensor3, tensor4]
d = {a: b for a, b in zip(inner_tensors, todo[-nb_tensor:])}  
todo = todo[nb_tensor:]
```
`nb_tensor` here is 2 . So basically d will be `{elem1: tensor3, elem2: tensor4}` but `todo[nb_tensor:]` will still give `tensor3, tensor4` which is wrong.

Differential Revision: [D66977210](https://our.internmc.facebook.com/intern/diff/D66977210)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66977210).

File: torch/_functorch/_aot_autograd/subclass_parametrization.py [hunk 1]
```diff
@@ -9,7 +9,7 @@ def forward(self, *tensors) -> torch.Tensor:  # type: ignore[no-untyped-def]
         todo: List[torch.Tensor] = list(tensors)
         for tp, meta, inner_tensors in reversed(self.rebuild_stack):
             nb_tensor: int = len(inner_tensors)
-            d = {a: b for a, b in zip(inner_tensors, todo[-nb_tensor:])}  # noqa: C416
+            d = {a: b for a, b in zip(inner_tensors, todo[:nb_tensor])}  # noqa: C416
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/subclass_parametrization.py:12]: cc @IvanKobzarev can you take a look too?

Comment (IvanKobzarev) [reply torch/_functorch/_aot_autograd/subclass_parametrization.py:12]: @bdhirsh @tugsbayasgalan 
Oh, Thanks Tugsu for finding this, that's a good one.
The problem seems to be deeper.
E.g. we have 
```
TwoTensor(TwoTensor(t1, t2), t0).
```
After right_inverse we have:
```
rebuilt_stack == [(TwoTensor, meta, ["a", "b"]), (TwoTensor, meta, ["a", "b"])]
plain_tensors == [t0, t1, t2]
```
We will first put plain tensors, and only then the nested TwoTensor.

But when we unflatten:
todo = [t0, t1, t2]
we first create TwoTensor[t1, t2]
 put it to todo [t0, TwoTensor[t1, t2]]
 And as a result get 
 TwoTensor(t0, TwoTensor(t1, t2)) which is swapping original a and b :)
So the fix should be different, we need to preserve the order of elements in the stack for plain/subclasses.
I will think about the fix.

Comment (IvanKobzarev) [reply torch/_functorch/_aot_autograd/subclass_parametrization.py:12]: The fix:

https://github.com/pytorch/pytorch/pull/142481

**Pull request closed without merging**